### PR TITLE
android: only fire a single 'terminal' callback per stream

### PR DESCRIPTION
**Pull request by goaway**
Description: Updates callback logic to ensure only one of any terminal callback is fired per stream/requests: one of final-onHeaders, final-onData, onTrailers, onError, and onCancel. Additionally ensures that it will be the last callback (iff provided callback dispatch is serial).
Risk: Low
Testing: Local and CI

Signed-off-by: Mike Schore <mike.schore@gmail.com>

### CODE REVIEW

**Comment by junr03** [comment]
related iOS change #610

File: library/java/src/io/envoyproxy/envoymobile/engine/JvmCallbackContext.java [hunk 1]
```diff
@@ -29,6 +29,20 @@
 
   public JvmCallbackContext(EnvoyHTTPCallbacks callbacks) { this.callbacks = callbacks; }
 
+  /**
+   * Return whether a callback should be allowed to continue with execution. This ensures at most
+   * one 'terminal' callback is issued for any given stream.
+   *
+   * @param close, whether the stream should be closed as part of this determination.
+   */
+  private dispatchable(boolean close) {
```
Comment (junr03) [inline library/java/src/io/envoyproxy/envoymobile/engine/JvmCallbackContext.java]: missing return type

File: library/java/src/io/envoyproxy/envoymobile/engine/JvmCallbackContext.java [hunk 2]
```diff
@@ -167,14 +181,16 @@ public void run() {
...
-   * @return boolean, whether the callback context was canceled or not.
+   * @return boolean, whether the callback context was closed or not.
    */
   public boolean cancel() {
-    boolean canceled = !this.canceled.getAndSet(true);
-    if (canceled) {
+    // Atomically close the stream if not already closed.
+    boolean closed = dispatchable(true);
+    if (closed) {
+      // Directly fire callback if closure occurred.
```
Comment (junr03) [inline library/java/src/io/envoyproxy/envoymobile/engine/JvmCallbackContext.java]: can you remind me why we don't reset the stream here?

File: library/java/src/io/envoyproxy/envoymobile/engine/JvmCallbackContext.java (hunk 1 above)
Comment (goaway) [reply library/java/src/io/envoyproxy/envoymobile/engine/JvmCallbackContext.java]: (already fixed)

File: library/java/src/io/envoyproxy/envoymobile/engine/JvmCallbackContext.java (hunk 2 above)
Comment (goaway) [reply library/java/src/io/envoyproxy/envoymobile/engine/JvmCallbackContext.java]: https://github.com/lyft/envoy-mobile/blob/ms/java-termcbs/library/java/src/io/envoyproxy/envoymobile/engine/EnvoyHTTPStream.java#L80

Comment (goaway) [reply library/java/src/io/envoyproxy/envoymobile/engine/JvmCallbackContext.java]: (but thank you for helping me catch a bug)

**Pull request merged**